db/sqldb_test.go

```diff
@@ -0,0 +1,245 @@
+		}
+	}
+}
+
+func TestCreate_trivial(t *testing.T) {
+	t.Parallel()
+	createSQL := `[INSERT INTO volumes(name) VALUES (?);]`
+
+	db, mock, volumeDatabase := createMockVolumeDatabase(t)
+	defer db.Close()
+
+	// Configure Mock
+	mock.ExpectBegin()
+	prepared := mock.ExpectPrepare(createSQL)
+	prepared.ExpectExec().WithArgs("volume_name").WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectCommit()
+
+	if err := volumeDatabase.Create("volume_name", map[string]string{}); err != nil {
+		t.Errorf("error was not expected while creating volume: %s", err)
+	}
+
+	// we make sure that all expectations were met
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("there were unfulfilled expections: %s", err)
+	}
+}
+
+func TestCreate_badNoName(t *testing.T) {
+	t.Parallel()
+
+	db, _, volumeDatabase := createMockVolumeDatabase(t)
+	defer db.Close()
+
+	err := volumeDatabase.Create("", map[string]string{})
+	if err == nil {
+		t.Error("error was expected while creating volume as name cannot be nil.")
+	}
+
+	if err.Error() != "volume name cannot be empty" {
+		t.Error("error msg was expected to be 'volume name cannot be empty', was '" + err.Error() + "'")
+	}
+}
+
+func TestCreate_failBegin(t *testing.T) {
+	t.Parallel()
+
+	db, mock, volumeDatabase := createMockVolumeDatabase(t)
+	defer db.Close()
+
+	mock.ExpectBegin().WillReturnError(errors.New("ExampleError"))
+
+	err := volumeDatabase.Create("volume_name", map[string]string{})
+	if err == nil {
+		t.Error("error was expected while creating volume as name cannot be nil.")
+	}
+
+	if err.Error() != "ExampleError" {
+		t.Error("error msg was expected to be 'ExampleError', was '" + err.Error() + "'")
+	}
+
+	// we make sure that all expectations were met
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("there were unfulfilled expections: %s", err)
+	}
+}
+
+func TestCreate_failPrepare(t *testing.T) {
+	t.Parallel()
+	createSQL := `[INSERT INTO volumes(name) VALUES (?);]`
+
+	db, mock, volumeDatabase := createMockVolumeDatabase(t)
+	defer db.Close()
+
+	mock.ExpectBegin()
+	mock.ExpectPrepare(createSQL).WillReturnError(errors.New("ExampleError"))
+	mock.ExpectRollback()
+
+	err := volumeDatabase.Create("volume_name", map[string]string{})
+	if err == nil {
+		t.Error("error was expected while creating volume as name cannot be nil.")
+	}
+
+	if err.Error() != "ExampleError" {
+		t.Error("error msg was expected to be 'ExampleError', was '" + err.Error() + "'")
+	}
+
+	// we make sure that all expectations were met
+	if err := mock.ExpectationsWereMet(); err != nil {
+		t.Errorf("there were unfulfilled expections: %s", err)
+	}
+}
+
+func TestCreate_failExec(t *testing.T) {
+	t.Parallel()
+	createSQL := `[INSERT INTO volumes(name) VALUES (?);]`
+
+	db, mock, volumeDatabase := createMockVolumeDatabase(t)
+	defer db.Close()
+
+	mock.ExpectBegin()
+	prepare := mock.ExpectPrepare(createSQL)
+	prepare.ExpectExec().WithArgs("volume_name").WillReturnError(errors.New("ExampleError"))
```

Comment:
I see now how this works. It looks like the `mock.ExpectationsWereMet()` call below does the error logging for expectations.